src/Serializers/Orleans.Serialization.Protobuf/ProtobufCodec.cs

```diff
@@ -0,0 +1,210 @@
+using Google.Protobuf;
+using Orleans.Serialization.Buffers;
+using Orleans.Serialization.Buffers.Adaptors;
+using Orleans.Serialization.Cloning;
+using Orleans.Serialization.Codecs;
+using Orleans.Serialization.Serializers;
+using Orleans.Serialization.WireProtocol;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Orleans.Serialization;
+
+[Alias(WellKnownAlias)]
+public sealed class ProtobufCodec : IGeneralizedCodec, IGeneralizedCopier, ITypeFilter
+{
+    public const string WellKnownAlias = "protobuf";
+
+    private static readonly ConcurrentDictionary<RuntimeTypeHandle, MessageParser> Parsers = new();
+
+    private readonly ICodecSelector[] _serializableTypeSelectors;
+    private readonly ICopierSelector[] _copyableTypeSelectors;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ProtobufCodec"/> class.
+    /// </summary>
+    /// <param name="serializableTypeSelectors">Filters used to indicate which types should be serialized by this codec.</param>
+    /// <param name="copyableTypeSelectors">Filters used to indicate which types should be copied by this codec.</param>
+    public ProtobufCodec(
+        IEnumerable<ICodecSelector> serializableTypeSelectors,
+        IEnumerable<ICopierSelector> copyableTypeSelectors)
+    {
+        _serializableTypeSelectors = serializableTypeSelectors.Where(t => string.Equals(t.CodecName, WellKnownAlias, StringComparison.Ordinal)).ToArray();
+        _copyableTypeSelectors = copyableTypeSelectors.Where(t => string.Equals(t.CopierName, WellKnownAlias, StringComparison.Ordinal)).ToArray();
+    }
+
+    public object DeepCopy(object input, CopyContext context)
+    {
+        if (!context.TryGetCopy(input, out object result))
+        {
+            dynamic dynamicSource = input;
+            result = dynamicSource.Clone();
+
+            context.RecordCopy(input, result);
```

Comment:
I generally don't like this approach. I don't think it will work as intended in most cases.
It would be safer to perform a serialization loop.

Reply:
Protobuf `IMessage<T>` implements an `IDeepCloneable<T>` - the generated code does perform deep copies, if we can get at the interface. That is why I implemented `ISpecializableCopier` instead.

Reply:
> That is why I implemented `ISpecializableCopier` instead.

I'm not seeing that in this PR, am I missing something?

I don't suppose there're open to having APIs that box to `object` for these cases.

Reply:
Sorry, I'm not the pr author. I was talking about my gist I linked to in the discussion that spurred this pr.

Reply:
Oh, right, my apologies. I suppose this `dynamic` approach is *ok* given it's just calling into APIs we know are available, although I am not a fan, performance-wise.

Reply:
@larsfjerm what do you think about taking the `ISpecializable*` route? It would mean that codecs are created up-front for each type and dynamic invoke/reflection is avoided after that, and no boxing would be required. It adds the cost of calling `MakeGenericType` instead

Reply:
I refactored the dynamic clone into a full byte serialization in my latest commit, but I will have a look at this over the weekend and give it a go 👍

Reply:
Ok, I'll hold off on merging it until then